Realm/Tests/KVOTests.mm

```diff
@@ -0,0 +1,1278 @@
+    RLMRealm *_observationRealm;
+    NSMutableArray *_notifications;
+
+public:
+    // construct a new recorder for the given `keyPath` on `obj`, using `observer`
+    // as the NSObject helper to actually add as an observer
+    KVORecorder(id observer, id obj, NSString *keyPath,
+                int options = NSKeyValueObservingOptionOld|NSKeyValueObservingOptionNew)
+    : _observer(observer)
+    , _obj([observer observableForObject:obj])
+    , _keyPath(keyPath)
+    , _mutationRealm([obj respondsToSelector:@selector(realm)] ? (RLMRealm *)[obj realm] : nil)
+    , _observationRealm([_obj respondsToSelector:@selector(realm)] ? (RLMRealm *)[_obj realm] : nil)
+    , _notifications([NSMutableArray new])
+    {
+        [_obj addObserver:observer forKeyPath:keyPath options:options context:this];
+    }
+
+    ~KVORecorder() {
+        id self = _observer;
+        @try {
+            [_obj removeObserver:_observer forKeyPath:_keyPath context:this];
+        }
+        @catch (NSException *e) {
+            XCTFail(@"%@", e.description);
+        }
+        XCTAssertEqual(0U, _notifications.count);
+    }
+
+    // record a single notification
+    void operator()(NSString *key, id obj, NSDictionary *changeDictionary) {
+        id self = _observer;
+        XCTAssertEqual(obj, _obj);
+        XCTAssertEqualObjects(key, _keyPath);
+        [_notifications addObject:changeDictionary.copy];
+    }
+
+    // ensure that the observed object is updated for any changes made to the
+    // object being mutated if they are different
+    void refresh() {
+        if (_mutationRealm != _observationRealm) {
+            [_mutationRealm commitWriteTransaction];
+            [_observationRealm refresh];
+            [_mutationRealm beginWriteTransaction];
+        }
+    }
+
+    NSDictionary *pop_front() {
+        NSDictionary *value = [_notifications firstObject];
+        if (value) {
+            [_notifications removeObjectAtIndex:0U];
+        }
+        return value;
+    }
+
+    NSUInteger size() const {
+        return _notifications.count;
+    }
+
+    bool empty() const {
+        return _notifications.count == 0;
+    }
+};
+
+// Assert that `recorder` has a notification at `index` and return it if so
+#define AssertNotification(recorder) ([&]{ \
+    (recorder).refresh(); \
+    NSDictionary *value = recorder.pop_front(); \
+    XCTAssertNotNil(value, @"Did not get a notification when expected"); \
+    return value; \
+})()
+
+// Validate that `recorder` has at least one notification, and that the first
+// notification is the expected one
+#define AssertChanged(recorder, from, to) do { \
+    if (NSDictionary *note = AssertNotification((recorder))) { \
+        XCTAssertEqualObjects(@(NSKeyValueChangeSetting), note[NSKeyValueChangeKindKey]); \
+        XCTAssertEqualObjects((from), note[NSKeyValueChangeOldKey]); \
+        XCTAssertEqualObjects((to), note[NSKeyValueChangeNewKey]); \
+    } \
+    else { \
+        return; \
```

Comment:
yes. once there's one missing notification generally everything after it fails in increasing odd ways resulting in lots of error spam that makes it harder to find the actual problem.